incubator/orm/genesis.go

```diff
@@ -0,0 +1,135 @@
+
+import (
+	"bytes"
+	"encoding/json"
+	"reflect"
+
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	"github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/gogo/protobuf/proto"
+)
+
+// Model defines the IO structure for table imports and exports
+type Model struct {
+	Key   []byte          `json:"key", yaml:"key"`
+	Value json.RawMessage `json:"value", yaml:"value"`
+}
+
+// TableExportable
+type TableExportable interface {
+	// Table returns the table to export
+	Table() Table
+}
+
+// SequenceExportable
+type SequenceExportable interface {
+	// Sequence returns the sequence to export
+	Sequence() Sequence
+}
+
+// ExportTableData returns a json encoded `[]Model` slice of all the data persisted in the table.
+// When the given table implements the `SequenceExportable` interface then it's current value
+// is returned as well or otherwise defaults to 0.
+func ExportTableData(ctx HasKVStore, t TableExportable) (json.RawMessage, uint64, error) {
+	enc := jsonpb.Marshaler{}
+	var r []Model
+	forEachInTable(ctx, t.Table(), func(rowID RowID, obj Persistent) error {
+		pbObj, ok := obj.(proto.Message)
+		if !ok {
+			return errors.Wrapf(ErrType, "not a proto message type: %T", pbObj)
+		}
+		var buf bytes.Buffer
+		err := enc.Marshal(&buf, pbObj)
+		if err != nil {
+			return errors.Wrap(err, "json encoding")
+		}
+		r = append(r, Model{Key: rowID, Value: buf.Bytes()})
+		return nil
+	})
+	var seqValue uint64
+	if st, ok := t.(SequenceExportable); ok {
+		seqValue = st.Sequence().CurVal(ctx)
+	}
+	b, err := json.Marshal(r)
+	return b, seqValue, err
+}
+
+// ImportTableData initializes a table and attached indexers from the given json encoded `[]Model`s.
+// The seqValue is optional and only used with tables that implement the `SequenceExportable` interface.
+func ImportTableData(ctx HasKVStore, t TableExportable, src json.RawMessage, seqValue uint64) error {
+	dec := json.NewDecoder(bytes.NewReader(src))
+	if _, err := dec.Token(); err != nil {
+		return errors.Wrap(err, "open bracket")
+	}
+	table := t.Table()
+	if err := clearAllInTable(ctx, table); err != nil {
+		return errors.Wrap(err, "clear old entries")
+	}
+	for dec.More() {
+		var m Model
+		if err := dec.Decode(&m); err != nil {
+			return errors.Wrap(err, "decode")
+		}
+		if err := putIntoTable(ctx, table, m); err != nil {
+			return errors.Wrap(err, "insert from genesis model")
+		}
+	}
+	if _, err := dec.Token(); err != nil {
+		return errors.Wrap(err, "closing bracket")
+	}
+	if st, ok := t.(SequenceExportable); ok {
+		if err := st.Sequence().InitVal(ctx, seqValue); err != nil {
+			return errors.Wrap(err, "sequence")
+		}
+	}
+	return nil
+}
+
+// forEachInTable iterates through all entries in the given table and calls the callback function.
+// Aborts on first error.
+func forEachInTable(ctx HasKVStore, table Table, f func(RowID, Persistent) error) error {
+	store := prefix.NewStore(ctx.KVStore(table.storeKey), []byte{table.prefix})
```

Comment:
Shouldn't we just be able to call `table.PrefixScan(ctx, nil, nil)` to iterate through all entries in the table? Looks like it does basically the same thing.

Reply:
This is correct but you would get a type safe iterator back. I thought this would be overhead when iterating over a large number of entries. But if it makes code more readable I can switch. 🤔 should have done benchmarks first....

Reply:
as discussed -> reuse existing impl

Reply:
pushed a change